llm-examples/mpt/mpt-7b/scripts/fine_tune_deepspeed.py

```diff
@@ -0,0 +1,261 @@
+os.environ['TRANSFORMERS_CACHE'] = '/local_disk0/hf'
+
+logger = logging.getLogger(__name__)
+
+INTRO_BLURB = "Below is an instruction that describes a task. Write a response that appropriately completes the request."
+INSTRUCTION_KEY = "### Instruction:"
+INPUT_KEY = "Input:"
+RESPONSE_KEY = "### Response:"
+PROMPT_NO_INPUT_FORMAT = """{intro}
+
+{instruction_key}
+{instruction}
+
+{response_key}""".format(
+  intro=INTRO_BLURB,
+  instruction_key=INSTRUCTION_KEY,
+  instruction="{instruction}",
+  response_key=RESPONSE_KEY,
+)
+
+PROMPT_WITH_INPUT_FORMAT = """{intro}
+
+{instruction_key}
+{instruction}
+
+{input_key}
+{input}
+
+{response_key}""".format(
+  intro=INTRO_BLURB,
+  instruction_key=INSTRUCTION_KEY,
+  instruction="{instruction}",
+  input_key=INPUT_KEY,
+  input="{input}",
+  response_key=RESPONSE_KEY,
+)
+
+
+ROOT_PATH = Path(__file__).parent.parent
+MODEL_PATH = 'mosaicml/mpt-7b'
+TOKENIZER_PATH = 'EleutherAI/gpt-neox-20b'
+DEFAULT_TRAINING_DATASET = "databricks/databricks-dolly-15k"
+CONFIG_PATH = "../../config/a10_config.json"
+LOCAL_OUTPUT_DIR = "/local_disk0/output"
+DEFAULT_SEED = 68
+
+
+def load_training_dataset(
+  tokenizer,
+  path_or_dataset: str = DEFAULT_TRAINING_DATASET,
+  seed: int = DEFAULT_SEED
+  ) -> Dataset:
+    logger.info(f"Loading dataset from {path_or_dataset}")
+    dataset = load_dataset(path_or_dataset)["train"]
+    logger.info(f"Found {dataset.num_rows} rows", )
+
+    def _reformat_data(rec):
+      instruction = rec["instruction"]
+      response = rec["response"]
+      context = rec.get("context")
+
+      if context:
+        questions = PROMPT_WITH_INPUT_FORMAT.format(instruction=instruction, input=context)
+      else:
+        questions = PROMPT_NO_INPUT_FORMAT.format(instruction=instruction)
+      return {"text": f"{{ 'prompt': {questions}, 'response': {response} }}"}
+    
+    dataset = dataset.map(_reformat_data)
+
+    def tokenize_function(allEntries):
+      return tokenizer(allEntries['text'], truncation=True, max_length=512)
+    
+    dataset = dataset.map(tokenize_function)
+    split_dataset = dataset.train_test_split(test_size=1000, seed=seed)
+    train_tokenized_dataset = split_dataset['train']
+    eval_tokenized_dataset = split_dataset['test']
+
+    return train_tokenized_dataset, eval_tokenized_dataset
+      
+
+
+def load_model(
+    pretrained_model_name_or_path: str = MODEL_PATH,
+) -> AutoModelForCausalLM:
+    logger.info(f"Loading model for {pretrained_model_name_or_path}")
+    config = AutoConfig.from_pretrained(pretrained_model_name_or_path,
+                                        trust_remote_code=True
+                                        )
+    config.attn_config['attn_impl'] = 'torch'
+    # config.init_device = 'cuda'
+    model_hidden_size = config.d_model
+
+    model = transformers.AutoModelForCausalLM.from_pretrained(
+      pretrained_model_name_or_path,
+      config=config,
+      torch_dtype=torch.bfloat16,
+      trust_remote_code=True
+    )
```

Comment:
Should we pin the revision here, since mosaic changes there huggingface hub repo fairly often?